### ISSUE TITLE
Improve automated tests

### ISSUE THREAD

**Issue by niconoe** [tdwg/camtrap-dp#173]
This is a request we just discussed with @peterdesmet, I plan to tackle that in the near future (can someone assign me?):

- Start with existing tests in https://github.com/tdwg/camtrap-dp/blob/main/tests/validate_schemas.py (this is run frequently via GH actions, each time a change is done to the camtrap-dp standard)
- Apparently, it already check 4 files are valid JSON (good!) and that they are a valid table schema (not sure exactly what that covers, but looks good)

Improvements needed:

- Add a third test that validates that the example camtrap-dp package (will be provided in #118) is still compliant with the (just updated) new version of the standard?
- Also validate (what? the example package?) against camtrap-dp profile. (is this with `frictionless.py validate`)?
- Make code a bit cleaner/shorter/more pythonic (+ docstrings to state what it does)
- Make the output (visible on GitHub actions) a bit nicer (more clear and readable)

**Comment by peterdesmet** [comment]
1. Regarding the check whether something is a valid Table Schema, can you check if you need `tableschema-py` or if you can do it with [`frictionless-py validate_schema`](https://framework.frictionlessdata.io/docs/guides/validation-guide#validating-a-schema)?  If possible with the latter, then we only use one, maintained library (`frictionless-py`).

2. Regarding a `validate_examples.py`: to avoid confusing users, the example will always point to a stable version of Camtrap DP (as real datasets should), so testing the example as a whole will _not_ test it against the latest changes on a branch. Luckily, we can overwrite that (here written in CLI, should become Python):

```bash
cd example
# Will have to look up if it is possible to validate metadata against external profile
frictionless validate deployments.csv --schema-sync --schema ../deployments-table-schema.json
frictionless validate multimedia.csv --schema-sync --schema ../multimedia-table-schema.json
frictionless validate observations.csv --schema-sync --schema ../observations-table-schema.json
```

3. Somewhat fancier stuff, but it might be good to test as well if the example `datapackage.json` references the latest version of Camtrap DP. Can be done by searching for the 4 occurrences of `https://raw.githubusercontent.com/tdwg/camtrap-dp/`, which should be followed by the latest tag name.

**Comment by peterdesmet** [comment]
@niconoe to validate the metadata of the package, you'll have to validate the whole package, but then select the metadata errors only: `Package.metadata_errors`. I'm not sure yet how you can validate against an external profile. Maybe by loading it and then updating the profile link?

**Comment by peterdesmet** [comment]
@niconoe The example is available in the [example branch](https://github.com/tdwg/camtrap-dp/tree/example). Might be good to add validation test on that branch as well.

**Comment by niconoe** [comment]
Apparently our own profile does not allow schema to be an URL: https://github.com/frictionlessdata/frictionless-py/issues/935

I haven't really jumped into JSON schema: @peterdesmet: do you know an easy way to solve that? Otherwise I'll investigate further!

Not sure I understand the issue fully: I'm also a bit surprised that our custom profile can overwrite the properties of `schema`, as it is defined by the Frictionless standard. A related question is: will the "validate" commands follow the URLs and interpret the content behind, or just consider it as a string in that case.

**Comment by peterdesmet** [comment]
Ok, I'm trying to debug by creating a `test.py` in the root, that has:

```python
import json
from pprint import pprint
from frictionless import validate_package

with open("example/datapackage.json") as json_file:
  descriptor_data = json.load(json_file)

  report = validate_package(descriptor_data)
  for err in report.errors:  # report.valid is False
    pprint(err)
```

I am currently getting:

```
{'code': 'scheme-error',
 'description': 'Data reading error because of incorrect scheme.',
 'message': 'The data source could not be successfully loaded: [Errno 2] No '
            "such file or directory: 'deployments.csv'",
 'name': 'Scheme Error',
 'note': "[Errno 2] No such file or directory: 'deployments.csv'",
 'tags': []}
```

Why can it not find deployments.csv? Is it looking for it from the root?

**Comment by peterdesmet** [comment]
Update, if I place my `test.py` in the `example` dir it works. It also doesn't generate the error you mentioned in frictionless 935.

**Comment by niconoe** [comment]
@peterdesmet: Do I understand correctly that you don't see any error in the console if you create a `test.py` file in the `example` dir, and execute it with something like:

```
$ cd example
$ python test.py
```

Can you tell me the packages installed on your environment (`pip freeze`) Python version in use (`Python -V`)?

With Python 3.8.6 and the last version of frictionless-py (4.20.2), I still get:

```
{'code': 'package-error',
 'description': 'A validation cannot be processed.',
 'message': 'The data package has an error: '
            '"\'https://raw.githubusercontent.com/tdwg/camtrap-dp/0.1.6/deployments-table-schema.json\' '
            'is not of type \'object\'" at "resources/0/schema" in metadata '
            'and at '
            '"allOf/0/properties/resources/items/properties/schema/type" in '
            'profile',
 'name': 'Package Error',
 'note': '"\'https://raw.githubusercontent.com/tdwg/camtrap-dp/0.1.6/deployments-table-schema.json\' '
         'is not of type \'object\'" at "resources/0/schema" in metadata and '
         'at "allOf/0/properties/resources/items/properties/schema/type" in '
         'profile',
 'tags': []}

```

Sorry to bother you, but it seems we encounter difficulties to get a consistent result, so I'd like to understand that better.

**Comment by peterdesmet** [comment]
> Do I understand correctly that you don't see any error in the console if you create a test.py file in the example dir, and execute it with something like:

Correct.

Pip freeze:

<details>
<pre>
alabaster==0.7.12
aniso8601==8.1.0
appnope==0.1.0
argcomplete==1.12.1
attrs==19.1.0
awscli==1.16.165
awsebcli==3.15.2
Babel==2.9.0
backcall==0.1.0
beautifulsoup4==4.6.3
bitarray==1.5.3
bleach==3.1.0
blessed==1.15.0
blinker==1.4
boto3==1.14.51
botocore==1.17.51
cached-property==1.5.1
cement==2.8.2
Cerberus==1.2
certifi==2020.12.5
chardet==3.0.4
click==7.1.2
click-default-group==1.2.2
colorama==0.4.4
commonmark==0.9.1
datapackage==1.15.1
decorator==4.4.0
defusedxml==0.6.0
docker==3.7.2
docker-compose==1.23.2
docker-pycreds==0.4.0
dockerpty==0.4.1
docopt==0.6.2
docutils==0.14
entrypoints==0.3
et-xmlfile==1.0.1
feedgenerator==1.9.1
Flask==1.1.2
Flask-RESTful==0.3.8
frictionless==4.12.4
-e git+https://github.com/timrobertson100/frictionless-transforms.git@1e2ed7585f1d1777672b94aa54b374af5fcabcd7#egg=frictionless_transforms
FrictionlessDarwinCore==1.0.0
future==0.16.0
goodtables==2.5.1
h5py @ file:///opt/concourse/worker/volumes/live/98bee3fd-2687-4017-5421-d742d1f18c35/volume/h5py_1593454156895/work
idna==2.7
ijson==3.1.1
imagesize==1.2.0
importlib-metadata==2.0.0
ipykernel==5.1.0
ipython==7.5.0
ipython_genutils==0.2.0
ipywidgets==7.4.2
isodate==0.6.0
itsdangerous==1.1.0
jdcal==1.4.1
jedi==0.13.3
Jinja2==2.10.1
jinjasql==0.1.8
jmespath==0.9.4
jsonlines==1.2.0
jsonpath-ng==1.5.2
jsonpointer==2.0
jsonschema==2.6.0
jupyter==1.0.0
jupyter-client==5.2.4
jupyter-console==6.0.0
jupyter-core==4.4.0
linear-tsv==1.1.0
Markdown==3.0.1
MarkupSafe==1.1.1
mistune==0.8.4
mkl-fft==1.0.12
mkl-random==1.0.2
nbconvert==5.5.0
nbformat==4.4.0
notebook==5.7.8
numpy==1.16.3
openpyxl==3.0.5
packaging==20.8
pandas==0.24.2
pandocfilters==1.4.2
parso==0.4.0
pathspec==0.5.9
pelican==4.0.0
petl==1.7.2
pexpect==4.7.0
pickleshare==0.7.5
ply==3.11
prometheus-client==0.6.0
prompt-toolkit==2.0.9
psycopg2==2.8.6
ptyprocess==0.6.0
pyaml==17.12.1
pyasn1==0.4.5
pybloom-live==3.0.0
Pygments==2.4.0
pyodbc==4.0.30
pyparsing==2.4.7
pyrsistent==0.14.11
python-dateutil==2.8.1
python-dwca-reader==0.13.0
python-slugify==4.0.1
pytz==2019.1
pywhip==0.3.2
PyYAML==5.3.1
pyzmq==18.0.0
qtconsole==4.4.4
recommonmark==0.6.0
requests==2.22.0
rfc3986==1.4.0
rfc3987==1.3.7
rsa==3.4.2
s3transfer==0.3.3
semantic-version==2.5.0
Send2Trash==1.5.0
shellingham==1.4.0
simpleeval==0.9.10
simplejson==3.17.2
six==1.12.0
snowballstemmer==2.0.0
Sphinx==3.3.1
sphinx-markdown-tables==0.0.15
sphinx-rtd-theme==0.5.0
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
SQLAlchemy==1.3.19
statistics==1.0.3.5
strict-rfc3339==0.7
stringcase==1.2.0
tableschema==1.19.3
tableschema-sql==1.3.1
tabulate==0.8.7
tabulator==1.52.3
termcolor==1.1.0
terminado==0.8.2
testpath==0.4.2
text-unidecode==1.3
texttable==0.9.1
tornado==6.0.2
traitlets==4.3.2
typer==0.3.2
unicodecsv==0.14.1
Unidecode==1.1.1
urllib3==1.24.3
validators==0.18.2
wcwidth==0.1.7
webencodings==0.5.1
websocket-client==0.56.0
Werkzeug==1.0.1
widgetsnbextension==3.4.2
xlrd==1.2.0
xlwt==1.3.0
xmltodict==0.12.0
zipp==3.3.0
</pre>
</details>

Python -V: `Python 3.7.3`

**Comment by niconoe** [comment]
Thanks @peterdesmet, at least we have consistent behaviour now. I think the next step is to confirm an update is needed to our own profile to allow objects there (as suggested by roll in https://github.com/frictionlessdata/frictionless-py/issues/935).

We should indeed into this together. Prior to that, I plan to read a bit more about JSON schemas, frictionless specs, ... so far I have some trouble wrapping my brain around some concepts.

**Comment by peterdesmet** [comment]
@niconoe the bug has been fixed in the Frictionless specs. If I run the example/test.py above it returns without error (using `frictionless==4.22.0`). I guess that means you can finish a PR?

**Comment by niconoe** [comment]
Great! Indeed, I'll go back to this issue very soon.

**Comment by niconoe** [comment]
Quick update: 
- validation against versioned schemas is now implemented. 
- While working on that, I also submitted an issue to frictionless about the problem you discovered on Nov 4: https://github.com/frictionlessdata/frictionless-py/issues/956 (which I also faced, and looks like a bug to me).
- I'll now go back to the validation against current schemas

**Comment by niconoe** [comment]
@peterdesmet : validation against current schemas is now implemented. I think we're close.

I guess you'll need to merge into `main` before we can test the thing fully (because GitHub actions that react to push events on this branch)

**Comment by peterdesmet** [comment]
Automated tests run without issues: https://github.com/tdwg/camtrap-dp/actions Thanks @niconoe!
